wrappers/rust/icicle-cuda-runtime/src/memory.rs

```diff
@@ -117,6 +134,66 @@ impl<T> HostSlice<T> {
         self.0
             .iter_mut()
     }
+
+    pub fn is_pinnable(&self) -> bool {
+        let pinnable = get_device_attribute(cudaDeviceAttr::cudaDevAttrHostRegisterSupported, 0).unwrap();
+        let lockable =
+            get_device_attribute(cudaDeviceAttr::cudaDevAttrPageableMemoryAccessUsesHostPageTables, 0).unwrap();
+
+        pinnable == 1 && lockable == 0
+    }
+
+    pub fn pin(&self, flags: CudaHostRegisterFlags) -> CudaResult<()> {
+        unsafe {
+            let ptr = self.as_ptr() as *mut c_void;
+            let flags_to_set = flags.bits();
+            cudaHostRegister(ptr, self.len(), flags_to_set as c_uint).wrap()
+        }
+    }
+
+    pub fn unpin(&self) -> CudaResult<()> {
+        unsafe {
+            let mut flags = 0;
+            let ptr = self.as_ptr() as *mut c_void;
+            cudaHostGetFlags(&mut flags, ptr).wrap()?;
+            cudaHostUnregister(ptr).wrap()
+        }
+    }
+
+    pub fn allocate_pinned(count: usize, flags: CudaHostAllocFlags) -> CudaResult<&'static mut Self> {
+        let size = count
+            .checked_mul(size_of::<T>())
+            .unwrap_or(0);
+        if size == 0 {
+            return Err(CudaError::cudaErrorMemoryAllocation); //TODO: only CUDA backend should return CudaError
+        }
+
+        let mut pinned_host_ptr = MaybeUninit::<*mut c_void>::uninit();
+
+        unsafe {
+            cudaHostAlloc(pinned_host_ptr.as_mut_ptr(), size, flags.bits).wrap()?;
+            let pinned_host_slice = from_raw_parts_mut(pinned_host_ptr.assume_init() as *mut T, count);
+            Ok(Self::from_mut_slice(pinned_host_slice))
+        }
+    }
+
+    pub fn free_pinned(&self) -> CudaResult<()> {
```

Comment:
Yes. Since its not possible (to my knowledge) to know whether the pinned memory was through `pin` or `allocate_pinned` and each one has their own specific free function that needs to be called

Reply:
can you store it as a field in the HostSlice and based on that call free_pinned() when dropped?

Reply:
actually you need to unpin in both cases right?

Reply:
> can you store it as a field in the HostSlice and based on that call free_pinned() when dropped?

This was my first thought but it is a breaking change which I was trying to avoid.

>Or is it not dropped since it's a view to memory that is owned by someone else?

HostSlice is a wrapper struct around already allocated host memory so yea owned by someone else. HostSlice can still be dropped but it doesn't drop the underlying slice.
 
>in that case the allocated_pinned() may be problematic

Maybe I missed something here, can you explain? Do you mean it will leak the static underlying slice?